provider/manifest/service.go

```diff
@@ -201,13 +225,60 @@ loop:
 			}
 
 		case req := <-s.mreqch:
+			// Check that the manifest is valid
 			if err := validation.ValidateManifest(req.value.Manifest); err != nil {
 				s.session.Log().Error("manifest validation failed",
 					"err", err, "deployment", req.value.Deployment)
 				req.ch <- err
 				break
 			}
 
+			// Check that hostnames are not in use
+			allHostnames := make([]string, 0)
+			var hostnameRequiredError error
+			for _, mgroup := range req.value.Manifest.GetGroups() {
+				for leaseID, groupName := range s.activeLeases {
+					// Only check leases with a matching deployment ID & group name
+					if req.value.Deployment.Equals(leaseID.DeploymentID()) && groupName == mgroup.GetName() {
+						allHostnames = append(allHostnames, util.AllHostnamesOfManifestGroup(mgroup)...)
+
+						// For each service that exposes via an Ingress, then require a hsotname
+						if s.config.HTTPServicesRequireAtLeastOneHost {
+							for _, service := range mgroup.Services {
+								for _, expose := range service.Expose {
+									if nil == hostnameRequiredError && util.ShouldBeIngress(expose) && len(expose.Hosts) == 0 {
+										hostnameRequiredError = fmt.Errorf("%w: service %q exposed on %d:%s must have a hostname", errManifestRejected, service.Name, util.ExposeExternalPort(expose), expose.Proto)
+									}
+								}
+							}
+						}
+
+					}
+				}
+			}
+
+			if nil != hostnameRequiredError {
+				req.ch <- hostnameRequiredError
+				break
+			}
+			// Run this is in a separate goroutine and pass the answer back along with the original request
+			go func(req manifestRequest) {
+				err := <-s.hostnameService.CanReserveHostnames(allHostnames)
+				pendingManifests <- pendingManifestRequest{
+					req:                  req,
+					hostnamesCheckResult: err,
+				}
+			}(req)
+
+		case pendingReq := <-pendingManifests:
+			req := pendingReq.req // Get original request
+
+			// Check that the hostnames are available
+			if pendingReq.hostnamesCheckResult != nil {
+				req.ch <- pendingReq.hostnamesCheckResult
+				break
+			}
+
```

Comment:
After my last refactor I have pushed out everything into the manifest manager, so the service isn't really handling anything anymore. This probably is easier to work with since everything happens in the manifest.

The code that matches up group names has been moved into it's own function and just returns immediately if  it finds a problem.